File: pkg/node/js/fs.js
// fs implements node:fs. The Go host exposes a small set of synchronous file
// primitives that return a JSON envelope ({ok, ...} or {ok:false, code, msg})
// and carry binary payloads as base64. This module builds the Node surface on
// top: the *Sync calls, the callback forms, and the fs.promises namespace.

__bento_defineModule("fs", function (module, exports, require) {
  "use strict";

  const { Buffer } = require("buffer");
  const path = require("path");

  function makeError(env, syscall, pathArg) {
    const err = new Error(
      (env.code || "Error") + ": " + (env.msg || "operation failed") +
      (pathArg ? ", " + syscall + " '" + pathArg + "'" : "")
    );
    err.code = env.code || "UNKNOWN";
    err.errno = env.errno || -1;
    err.syscall = syscall;
    if (pathArg) err.path = pathArg;
    return err;
  }

  function call(name, args, syscall, pathArg) {
    const env = JSON.parse(name.apply(null, args));
    if (!env.ok) throw makeError(env, syscall, pathArg);
    return env;
  }

  function decode(env, encoding) {
    const buf = Buffer.from(env.b64 || "", "base64");
    if (!encoding || encoding === "buffer" || (encoding && encoding.encoding === "buffer")) {
      if (encoding && typeof encoding === "object" && encoding.encoding && encoding.encoding !== "buffer") {
        return buf.toString(encoding.encoding);
      }
      return buf;
    }
    const enc = typeof encoding === "string" ? encoding : encoding.encoding;
    return enc ? buf.toString(enc) : buf;
  }

  function toBuffer(data, encoding) {
    if (Buffer.isBuffer(data)) return data;
    if (data instanceof Uint8Array) return Buffer.from(data);
    if (typeof data === "string") return Buffer.from(data, encoding && encoding.encoding ? encoding.encoding : encoding || "utf8");
    return Buffer.from(String(data), "utf8");
  }

  function makeStats(s) {
    const stat = {
      dev: s.dev, ino: s.ino, mode: s.mode, nlink: s.nlink, uid: s.uid, gid: s.gid,
      rdev: s.rdev || 0, size: s.size, blksize: s.blksize || 4096, blocks: s.blocks || 0,
      atimeMs: s.atimeMs, mtimeMs: s.mtimeMs, ctimeMs: s.ctimeMs, birthtimeMs: s.birthtimeMs || s.ctimeMs,
      atime: new Date(s.atimeMs), mtime: new Date(s.mtimeMs), ctime: new Date(s.ctimeMs),
      birthtime: new Date(s.birthtimeMs || s.ctimeMs),
      isFile: () => s.kind === "file",
      isDirectory: () => s.kind === "dir",
      isSymbolicLink: () => s.kind === "symlink",
      isBlockDevice: () => false,
      isCharacterDevice: () => false,
      isFIFO: () => false,
      isSocket: () => false,
    };
    return stat;
  }

  // ---- synchronous API --------------------------------------------------
  function readFileSync(p, options) {
    const env = call(__bento_fs_read, [String(p)], "open", p);
    return decode(env, options);
  }
  function writeFileSync(p, data, options) {
    const buf = toBuffer(data, options);
    call(__bento_fs_write, [String(p), buf.toString("base64"), "w"], "open", p);
  }
  function appendFileSync(p, data, options) {
    const buf = toBuffer(data, options);
    call(__bento_fs_write, [String(p), buf.toString("base64"), "a"], "open", p);
  }
  function existsSync(p) {
    try { const env = JSON.parse(__bento_fs_stat(String(p))); return !!env.ok; }
    catch (e) { return false; }
  }
  function statSync(p) {
    return makeStats(call(__bento_fs_stat, [String(p)], "stat", p).stat);
  }
  function lstatSync(p) {
    return makeStats(call(__bento_fs_lstat, [String(p)], "lstat", p).stat);
  }
  function mkdirSync(p, options) {
    const recursive = options === true || (options && options.recursive);
    call(__bento_fs_mkdir, [String(p), !!recursive], "mkdir", p);
  }
  function rmdirSync(p, options) {
    const recursive = options && options.recursive;
    call(__bento_fs_rm, [String(p), !!recursive, false], "rmdir", p);
  }
  function rmSync(p, options) {
    const recursive = options && options.recursive;
    const force = options && options.force;
    try { call(__bento_fs_rm, [String(p), !!recursive, false], "unlink", p); }
    catch (e) { if (!force) throw e; }
  }
  function unlinkSync(p) {
    call(__bento_fs_rm, [String(p), false, false], "unlink", p);
  }
  function readdirSync(p, options) {
    const env = call(__bento_fs_readdir, [String(p)], "scandir", p);
    const withTypes = options && options.withFileTypes;
    if (withTypes) {
      return env.entries.map((e) => new Dirent(e.name, e.kind));
    }
    return env.entries.map((e) => e.name);
  }
  function renameSync(from, to) {
    call(__bento_fs_rename, [String(from), String(to)], "rename", from);
  }
  function copyFileSync(from, to) {
    call(__bento_fs_copy, [String(from), String(to)], "copyfile", from);
  }
  function realpathSync(p) {
    return call(__bento_fs_realpath, [String(p)], "realpath", p).path;
  }
  function readlinkSync(p) {
    return call(__bento_fs_readlink, [String(p)], "readlink", p).path;
  }
  function symlinkSync(target, p) {
    call(__bento_fs_symlink, [String(target), String(p)], "symlink", p);
  }
  function chmodSync(p, mode) {
    call(__bento_fs_chmod, [String(p), mode | 0], "chmod", p);
  }
  function mkdtempSync(prefix) {
    return call(__bento_fs_mkdtemp, [String(prefix)], "mkdtemp", prefix).path;
  }

  class Dirent {
    constructor(name, kind) { this.name = name; this._kind = kind; }
    isFile() { return this._kind === "file"; }
    isDirectory() { return this._kind === "dir"; }
    isSymbolicLink() { return this._kind === "symlink"; }
    isBlockDevice() { return false; }
    isCharacterDevice() { return false; }
    isFIFO() { return false; }
    isSocket() { return false; }
  }

  // ---- callback API -----------------------------------------------------
  // Each async call runs its synchronous core on a microtask and dispatches the
  // node-style (err, result) callback. True nonblocking I/O lands in a later
  // milestone; the observable contract (async delivery, error-first) holds now.
  function asyncify(syncFn) {
    return function () {
      const args = Array.prototype.slice.call(arguments);
      const cb = args.pop();
      if (typeof cb !== "function") throw new TypeError("callback must be a function");
      Promise.resolve().then(() => {
        let result, error;
        try { result = syncFn.apply(null, args); }
        catch (e) { error = e; }
        if (error) cb(error);
        else if (result === undefined) cb(null);
        else cb(null, result);
      });
    };
  }

  const promises = {};
  function promisifySync(syncFn) {
    return function () {
      const args = arguments;
      return new Promise((resolve, reject) => {
        Promise.resolve().then(() => {
          try { resolve(syncFn.apply(null, args)); }
          catch (e) { reject(e); }
        });
      });
    };
  }

  const api = {
    readFileSync: readFileSync,
    writeFileSync: writeFileSync,
    appendFileSync: appendFileSync,
    existsSync: existsSync,
    statSync: statSync,
    lstatSync: lstatSync,
    mkdirSync: mkdirSync,
    rmdirSync: rmdirSync,
    rmSync: rmSync,
    unlinkSync: unlinkSync,
    readdirSync: readdirSync,
    renameSync: renameSync,
    copyFileSync: copyFileSync,
    realpathSync: realpathSync,
    readlinkSync: readlinkSync,
    symlinkSync: symlinkSync,
    chmodSync: chmodSync,
    mkdtempSync: mkdtempSync,
    Dirent: Dirent,
    constants: {
      F_OK: 0, R_OK: 4, W_OK: 2, X_OK: 1,
      O_RDONLY: 0, O_WRONLY: 1, O_RDWR: 2, O_CREAT: 64, O_EXCL: 128, O_TRUNC: 512, O_APPEND: 1024,
    },
    accessSync: function (p) { statSync(p); },
    access: undefined,
  };

  // Wire the callback and promise forms for the file operations.
  const asyncNames = {
    readFile: readFileSync, writeFile: writeFileSync, appendFile: appendFileSync,
    stat: statSync, lstat: lstatSync, mkdir: mkdirSync, rmdir: rmdirSync, rm: rmSync,
    unlink: unlinkSync, readdir: readdirSync, rename: renameSync, copyFile: copyFileSync,
    realpath: realpathSync, readlink: readlinkSync, symlink: symlinkSync, chmod: chmodSync,
    mkdtemp: mkdtempSync,
  };
  for (const name of Object.keys(asyncNames)) {
    api[name] = asyncify(asyncNames[name]);
    promises[name] = promisifySync(asyncNames[name]);
  }
  api.access = function (p, mode, cb) {
    if (typeof mode === "function") { cb = mode; }
    asyncify(statSync)(p, cb);
  };
  promises.access = promisifySync(statSync);

  api.promises = promises;
  module.exports = api;
});

File: pkg/node/fs.go
package node

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/tamnd/bento/pkg/nodehost"
)

// fsResult is the JSON envelope every filesystem host function returns. OK
// distinguishes success from a mapped error; the remaining fields are populated
// per operation and omitted when empty so the JS layer sees a tidy object.
type fsResult struct {
	OK bool `json:"ok"`
	// Code, Errno and Desc are the three parts of a Node filesystem error: the
	// string a program branches on, the number err.errno carries, and libuv's
	// description, which the JavaScript side puts between the code and the syscall
	// to build the message Node builds.
	Code  string `json:"code,omitempty"`
	Errno int    `json:"errno,omitempty"`
	Desc  string `json:"desc,omitempty"`
	// Syscall is set only when the call failed at a later step than the one the
	// JavaScript side would report on its own, and it overrides that name; see
	// failAt.
	Syscall string     `json:"syscall,omitempty"`
	B64     string     `json:"b64,omitempty"`
	Path    string     `json:"path,omitempty"`
	Stat    *statInfo  `json:"stat,omitempty"`
	Entries []dirEntry `json:"entries,omitempty"`
}

// statInfo is the platform-neutral stat snapshot the fs module turns into a
// Node Stats object. Times are milliseconds since the Unix epoch.
type statInfo struct {
	Size    int64   `json:"size"`
	Mode    uint32  `json:"mode"`
	Kind    string  `json:"kind"`
	MtimeMs float64 `json:"mtimeMs"`
	AtimeMs float64 `json:"atimeMs"`
	CtimeMs float64 `json:"ctimeMs"`
	Dev     int     `json:"dev"`
	Ino     int     `json:"ino"`
	Nlink   int     `json:"nlink"`
	UID     int     `json:"uid"`
	GID     int     `json:"gid"`
}

// dirEntry names one child of a directory and its kind (file, dir, symlink).
type dirEntry struct {
	Name string `json:"name"`
	Kind string `json:"kind"`
}

// jsonString marshals any envelope to a string for return across the bridge.
func jsonString[T any](v T) string {
	b, err := json.Marshal(v)
	if err != nil {
		return `{"ok":false,"code":"UNKNOWN","desc":"marshal failed"}`
	}
	return string(b)
}

// ok builds a success envelope; callers set the payload fields they need.
func ok(r fsResult) string {
	r.OK = true
	return jsonString(r)
}

// fail maps a Go error to a Node-style error envelope: the code a program
// branches on, the number err.errno carries, and libuv's description of it, which
// the JavaScript side assembles into Node's message. The classification is
// nodehost's so the interpreter and the AOT path answer the same thing, and so
// the Windows translation lives in one place; see pkg/nodehost/fserror.go for why
// it cannot be a comparison against syscall.ENOENT and friends.
func fail(err error) string {
	e := nodehost.ClassifyFSError(err)
	return jsonString(fsResult{OK: false, Code: e.Code, Errno: e.Errno, Desc: e.Desc})
}

// failAt is fail for a call that got further than the JavaScript side assumes.
// readFileSync reports itself as the open, since that is the step that fails for
// a path that is missing or unreadable, but a directory opens perfectly well and
// fails on the read. Node names the step that actually failed and gives no path
// at all in that case, printing "EISDIR: illegal operation on a directory, read",
// so the host says which syscall it was and the JS side takes its word for it.
func failAt(err error, syscall string) string {
	e := nodehost.ClassifyFSError(err)
	return jsonString(fsResult{OK: false, Code: e.Code, Errno: e.Errno, Desc: e.Desc, Syscall: syscall})
}

// fsHostFuncs returns the synchronous filesystem primitives the fs module builds
// on. The fs.promises and callback forms in JavaScript are layered on these.
func fsHostFuncs() map[string]HostFunc {
	return map[string]HostFunc{
		"__bento_fs_read":     hostFSRead,
		"__bento_fs_write":    hostFSWrite,
		"__bento_fs_stat":     func(a []any) (any, error) { return statEnvelope(str(a, 0), false), nil },
		"__bento_fs_lstat":    func(a []any) (any, error) { return statEnvelope(str(a, 0), true), nil },
		"__bento_fs_mkdir":    hostFSMkdir,
		"__bento_fs_rm":       hostFSRm,
		"__bento_fs_readdir":  hostFSReaddir,
		"__bento_fs_rename":   hostFSRename,
		"__bento_fs_copy":     hostFSCopy,
		"__bento_fs_realpath": hostFSRealpath,
		"__bento_fs_readlink": hostFSReadlink,
		"__bento_fs_symlink":  hostFSSymlink,
		"__bento_fs_chmod":    hostFSChmod,
		"__bento_fs_mkdtemp":  hostFSMkdtemp,
	}
}

func hostFSRead(a []any) (any, error) {
	data, err := os.ReadFile(str(a, 0))
	if err != nil {
		// os.ReadFile opens and then reads, and it says which of the two failed in
		// the PathError's Op, so a directory comes back as a failed read and is
		// reported as one.
		var perr *fs.PathError
		if errors.As(err, &perr) && perr.Op == "read" {
			return failAt(err, "read"), nil
		}
		return fail(err), nil
	}
	return ok(fsResult{B64: base64.StdEncoding.EncodeToString(data)}), nil
}

func hostFSWrite(a []any) (any, error) {
	data, decErr := base64.StdEncoding.DecodeString(str(a, 1))
	if decErr != nil {
		return fail(decErr), nil
	}
	flag := os.O_CREATE | os.O_WRONLY
	if str(a, 2) == "a" {
		flag |= os.O_APPEND
	} else {
		flag |= os.O_TRUNC
	}
	f, err := os.OpenFile(str(a, 0), flag, 0o644)
	if err != nil {
		return fail(err), nil
	}
	_, werr := f.Write(data)
	cerr := f.Close()
	if werr != nil {
		return fail(werr), nil
	}
	if cerr != nil {
		return fail(cerr), nil
	}
	return ok(fsResult{}), nil
}

func statEnvelope(path string, link bool) string {
	info, err := os.Stat(path)
	if link {
		info, err = os.Lstat(path)
	}
	if err != nil {
		return fail(err)
	}
	return ok(fsResult{Stat: statOf(info)})
}

func statOf(info os.FileInfo) *statInfo {
	kind := "file"
	switch {
	case info.IsDir():
		kind = "dir"
	case info.Mode()&fs.ModeSymlink != 0:
		kind = "symlink"
	}
	mtimeMs := float64(info.ModTime().UnixNano()) / 1e6
	return &statInfo{
		Size:    info.Size(),
		Mode:    uint32(info.Mode().Perm()),
		Kind:    kind,
		MtimeMs: mtimeMs,
		AtimeMs: mtimeMs,
		CtimeMs: mtimeMs,
		Nlink:   1,
	}
}

func hostFSMkdir(a []any) (any, error) {
	path := str(a, 0)
	var err error
	if boolArg(a, 1) {
		err = os.MkdirAll(path, 0o755)
	} else {
		err = os.Mkdir(path, 0o755)
	}
	if err != nil {
		return fail(err), nil
	}
	return ok(fsResult{Path: path}), nil
}

func hostFSRm(a []any) (any, error) {
	path := str(a, 0)
	var err error
	if boolArg(a, 1) {
		err = os.RemoveAll(path)
	} else {
		err = os.Remove(path)
	}
	if err != nil {
		return fail(err), nil
	}
	return ok(fsResult{}), nil
}

func hostFSReaddir(a []any) (any, error) {
	entries, err := os.ReadDir(str(a, 0))
	if err != nil {
		return fail(err), nil
	}
	list := make([]dirEntry, 0, len(entries))
	for _, e := range entries {
		kind := "file"
		switch {
		case e.IsDir():
			kind = "dir"
		case e.Type()&fs.ModeSymlink != 0:
			kind = "symlink"
		}
		list = append(list, dirEntry{Name: e.Name(), Kind: kind})
	}
	return ok(fsResult{Entries: list}), nil
}

func hostFSRename(a []any) (any, error) {
	if err := os.Rename(str(a, 0), str(a, 1)); err != nil {
		return fail(err), nil
	}
	return ok(fsResult{}), nil
}

func hostFSCopy(a []any) (any, error) {
	data, err := os.ReadFile(str(a, 0))
	if err != nil {
		return fail(err), nil
	}
	if err := os.WriteFile(str(a, 1), data, 0o644); err != nil {
		return fail(err), nil
	}
	return ok(fsResult{}), nil
}

func hostFSRealpath(a []any) (any, error) {
	p, err := filepath.Abs(str(a, 0))
	if err != nil {
		return fail(err), nil
	}
	// Node's realpathSync throws when the path does not exist rather than
	// answering with a path that names nothing, and every other reason
	// EvalSymlinks fails, a symlink loop or a permission denial, is a throw there
	// too. Swallowing the error handed the caller a path it had no evidence for.
	resolved, rerr := filepath.EvalSymlinks(p)
	if rerr != nil {
		return fail(rerr), nil
	}
	return ok(fsResult{Path: resolved}), nil
}

func hostFSReadlink(a []any) (any, error) {
	target, err := os.Readlink(str(a, 0))
	if err != nil {
		return fail(err), nil
	}
	return ok(fsResult{Path: target}), nil
}

func hostFSSymlink(a []any) (any, error) {
	if err := os.Symlink(str(a, 0), str(a, 1)); err != nil {
		return fail(err), nil
	}
	return ok(fsResult{}), nil
}

func hostFSChmod(a []any) (any, error) {
	if err := os.Chmod(str(a, 0), os.FileMode(intArg(a, 1)&0o777)); err != nil {
		return fail(err), nil
	}
	return ok(fsResult{}), nil
}

func hostFSMkdtemp(a []any) (any, error) {
	// Node appends six random characters directly to the prefix in its parent
	// directory. MkdirTemp substitutes the trailing "*" with the random run.
	//
	// The split is filepath.Split and not Dir plus Base, because those two do not
	// compose back to the prefix when it ends in a separator: Dir("/tmp/") is
	// "/tmp" and Base("/tmp/") is "tmp", so mkdtempSync("/tmp/") would have made
	// /tmp/tmpXXXXXX where Node makes /tmp/XXXXXX. Split leaves the stem empty,
	// which with the appended "*" says exactly what the caller meant.
	prefix := str(a, 0)
	parent, stem := filepath.Split(prefix)
	if parent == "" {
		parent = "."
	}
	dir, err := os.MkdirTemp(parent, stem+"*")
	if err != nil {
		return fail(err), nil
	}
	return ok(fsResult{Path: dir}), nil
}
